File: iosApp/Shared/TrackingCommand.swift
// Z.5d: the watchOS analogue of Android's `TrackingCommand`/`TrackingCommandCodec` (see
// `core/data/src/commonMain/kotlin/com/mileway/core/data/watch/TrackingCommand.kt`) — a
// watch-initiated start/stop message, mirroring `MilewaySyncModels.swift`'s pattern of a plain
// `Codable` struct shared between the `iosApp` and `MilewayWatch` targets (see `iosApp/project.yml`)
// so neither side needs the other's KMP framework to agree on wire shape.
//
// Deliberately its own tiny type (not reusing `MilewaySyncPayload`, which flows phone->watch and
// carries snapshot data, not a command) — same reasoning `TrackingCommand.kt` gives for not reusing
// `WatchSyncPayload` on the Android side.
import Foundation

struct TrackingCommand: Codable, Equatable {
    enum Action: String, Codable {
        case start
        case stop
    }

    var action: Action

    static let messageKey = "action"
}

/// `WCSession.sendMessage`/`didReceiveMessage` traffic in plain `[String: Any]` dictionaries, same
/// as `MilewaySyncCodec` for `updateApplicationContext` — this mirrors that codec's shape/fail-soft
/// contract for the message-based command channel.
enum TrackingCommandCodec {
    static func encode(_ command: TrackingCommand) -> [String: Any] {
        [TrackingCommand.messageKey: command.action.rawValue]
    }

    /// Returns `nil` (rather than throwing) on any malformed/unrecognized message, mirroring
    /// Android's `TrackingCommandCodec.decode` fail-soft contract.
    static func decode(_ message: [String: Any]) -> TrackingCommand? {
        guard
            let raw = message[TrackingCommand.messageKey] as? String,
            let action = TrackingCommand.Action(rawValue: raw)
        else {
            return nil
        }
        return TrackingCommand(action: action)
    }
}

File: core/data/src/commonMain/kotlin/com/mileway/core/data/watch/TrackingCommand.kt
package com.mileway.core.data.watch

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * P2.10: the watch->phone half of the peer-sync story (P2.9 is phone->watch). A watch-initiated
 * start/stop command, carrying the session token the phone's `WearTrackingCommandService`
 * (`app/src/gms`) needs to call `TrackingController.start`/`stop` with — the same token shape
 * `feature:tracking`'s `TrackingController` already uses (see
 * `feature.tracking.manager.TrackingController`; `core:data` cannot depend on `feature:tracking`,
 * so this is a plain token string, not a reference to that interface).
 *
 * Deliberately its own tiny wire type (not a reuse of [WatchSyncPayload], which flows the other
 * direction and carries snapshot data, not a command) so the two directions stay independently
 * versionable.
 */
@Serializable
data class TrackingCommand(
    val action: Action,
    val token: String,
) {
    @Serializable
    enum class Action { START, STOP }
}

/**
 * The JSON codec [TrackingCommand] transports travel over (Wear OS `MessageClient` bytes today;
 * any future watchOS `WCSession` message would reuse the same shape). Factored out exactly like
 * [SnapshotCacheCodec] so it is unit-testable without a platform-backed `MessageClient`.
 */
object TrackingCommandCodec {
    private val json = Json { ignoreUnknownKeys = true }

    fun encode(command: TrackingCommand): ByteArray {
        val encoded = json.encodeToString(TrackingCommand.serializer(), command)
        return encoded.encodeToByteArray()
    }

    /** Returns null (rather than throwing) on any malformed/legacy-shape message payload. */
    fun decode(bytes: ByteArray?): TrackingCommand? {
        if (bytes == null || bytes.isEmpty()) return null
        return try {
            json.decodeFromString(TrackingCommand.serializer(), bytes.decodeToString())
        } catch (_: Exception) {
            null
        }
    }
}
